### PR TITLE
Remove the local polymorphic flag hack.

### PR DESCRIPTION
**Pull request by maximedenes**
Some code in typeclasses was even breaking the invariant that
use_polymorphic_flag should not be called twice, but that code was
morally dead it seems, so we remove it.

### CODE REVIEW

**Comment by maximedenes** [comment]
I copied the logic of `Set Program Mode` before realizing that it doesn't seem to work at all, so this is probably going to require some fixing.

**Comment by maximedenes** [comment]
Ok, I fixed the condition so there is some hope.

File: vernac/classes.ml [hunk 1]
```diff
@@ -401,8 +400,7 @@ let context poly l =
       let cst = Declare.declare_constant ~internal:Declare.InternalTacticRequest id decl in
 	match class_of_constr !evars (EConstr.of_constr t) with
 	| Some (rels, ((tc,_), args) as _cl) ->
-	    add_instance (Typeclasses.new_instance tc Hints.empty_hint_info false (*FIXME*)
```
Comment (SkySkimmer) [inline vernac/classes.ml]: Was this FIXME for poly stuff?

File: vernac/vernacentries.ml [hunk 2]
```diff
@@ -2195,6 +2191,7 @@ let with_fail st b f =
   end
 
 let interp ?(verbosely=true) ?proof ~st (loc,c) =
+  let orig_univ_poly = Flags.is_universe_polymorphism () in
   let orig_program_mode = Flags.is_program_mode () in
   let rec aux ?polymorphism ~atts isprogcmd = function
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: Should `polymorphism` go in the `atts`?

File: vernac/classes.ml (hunk 1 above)
Comment (maximedenes) [reply vernac/classes.ml]: Good question!

File: vernac/vernacentries.ml (hunk 2 above)
Comment (maximedenes) [reply vernac/vernacentries.ml]: Note quite. IIUC, `polymorphism` is the value written explicitly by the user (a `bool option`). In `atts`, you find the computed value (a `bool`).

Comment (maximedenes) [reply vernac/vernacentries.ml]: The distinction is "useful" to make sure polymorphism is not specified twice. @vbgl is probably going to get rid of this redundancy in his attribute branch.

**Review by ejgallego — APPROVED** [review]
Another very good step towards cleaning up command interp. Good to go if testing works.

**Comment by ejgallego** [comment]
@maximedenes , in my split_command branch I'm down to 11 `Flags.program_mode` uses in `pretyping` [4 in `cases`, 4 in `coercion`, and 3 in `pretyping` and 1 in `evarutil`.

Would we manage to solve the lowest ones we could also get rid of that flag completely. [c.f. #6216 ]

With this patch, the `Flags.universe_polymorphics` looks almost dead! Yay!!

**Comment by maximedenes** [comment]
Yeah I'm removing other occurrences of `Flags.universe_polymorphic`, see #6401 and upcoming PRs.

**Comment by maximedenes** [comment]
It seems one occurrence is going to stay, though, see the one show in #6403. At least until we can thread attributes all the way to the printer...

**Comment by SkySkimmer** [comment]
Why the needs: review given the approved from @ejgallego?

**Comment by ejgallego** [comment]
> Why the needs: review given the approved from @ejgallego?

He's sabotaging his own PRs so he can claim he's the stricter on them XDD

**Comment by maximedenes** [comment]
> Why the needs: review given the approved from @ejgallego?

I probably overlooked the approval. Sorry about that!

File: vernac/classes.ml (hunk 1 above)
Comment (maximedenes) [reply vernac/classes.ml]: I think so, because it appeared in several of these functions.

**Pull request merged**